app/controllers/payments_controller.rb

```diff
@@ -95,52 +162,61 @@ def webhook
 
-  def error
-    helpers.flash_message(:alert, t('.error'))
-    payment = Payment.find(params[:id])
-    redirect_on_payment_success_or_error(payment)
-  end
+  def handle_order_confirmation(klarna_id, payment_id)
 
-  private
+    klarna_order = KlarnaService.get_checkout_order(klarna_id)
 
-  def redirect_on_payment_success_or_error(payment)
+    KlarnaService.acknowledge_order(klarna_id)
 
-    if payment.payment_type == Payment::PAYMENT_TYPE_MEMBER
-      redirect_to user_path(params[:user_id])
-    else
-      redirect_to company_path(payment.company_id)
-    end
+    payment = Payment.find(payment_id)
+    payment.update_attribute(:status,
+                             Payment.order_to_payment_status(klarna_order['status']))
+    payment.successfully_completed
+
+    # Capture the order in Klarna (indicates the order has been filled and
+    # payment settlement can occur)
+    KlarnaService.capture_order(klarna_id, klarna_order['order_amount'])
+
+    klarna_order
   end
 
-  def log_hips_activity(activity, severity, payment_id, hips_id, exc=nil)
-    ActivityLogger.open(HIPS_LOG, 'HIPS_API', activity, false) do |log|
+  def log_klarna_activity(activity, severity, payment_id, klarna_id, exc=nil)
+    ActivityLogger.open(KLARNA_LOG, 'Payments', activity, false) do |log|
       log.record(severity, "Payment ID: #{payment_id}") if payment_id
-      log.record(severity, "HIPS ID: #{hips_id}") if hips_id
-      log.record(severity, "Exception class: #{exc.class}") if exc
-      log.record(severity, "Exception message: #{exc.message}") if exc
+      log.record(severity, "Klarna ID: #{klarna_id}") if klarna_id
+
+      [exc, exc&.cause].each do |this_exc|
+        next unless this_exc
+        log.record(severity, "Exception class: #{this_exc.class}") if exc
+        log.record(severity, "Exception message: #{this_exc.message}") if exc
+      end
     end
   end
 
-  def hips_order_urls(user_id, payment_id, company_id, payment_type)
+  def klarna_order_urls(payment_id, user_id, company_id, payment_type)
     urls = {}
-    urls[:success] = payment_success_url(user_id: user_id,
-                                         company_id: company_id,
-                                         id: payment_id,
-                                         payment_type: payment_type)
+    urls[:checkout] = payments_url(user_id: user_id, company_id: company_id,
+                                   type: payment_type)
 
-    urls[:error]   = payment_error_url(user_id: user_id,
-                                       company_id: company_id,
-                                       id: payment_id,
-                                       payment_type: payment_type)
+    urls[:confirmation] = payment_confirmation_url(id: payment_id,
+                                                   user_id: user_id,
+                                                   disable_language_change: true,
+                                                   klarna_id: '{checkout.order.id}')
```

Comment:
No ... Klarna will replace that string with actual ID when that webhook URL is invoked.

Reply:
Can you put that in a comment, please?  I won't be the only one to wonder that when looking at the code. :-)